h2o-core/src/main/java/water/rapids/PermutationVarImp.java

```diff
@@ -0,0 +1,217 @@
+     * sets the allowed metrics
+     *
+     * @param model trained model
+     * @param fr    training frame
+     */
+    public PermutationVarImp(Model model, Frame fr) {
+        if (fr.numRows() < 2)
+            throw new IllegalArgumentException("Frame must contain more than 1 rows to be used in permutation variable importance!");
+        _model = model;
+        _inputFrame = fr;
+    }
+
+    /**
+     * Returns the metric (loss function) selected by the user (mse is default)
+     *
+     * @throws IllegalArgumentException if metric could not be loaded
+     */
+    private static double getMetric(ModelMetrics mm, String metric) {
+        assert mm != null;
+        double metricValue = ModelMetrics.getMetricFromModelMetric(mm, metric);
+        if (Double.isNaN(metricValue))
+            throw new IllegalArgumentException("Model doesn't support the metric following metric " + metric);
+        return metricValue;
+    }
+
+    private String inferAndValidateMetric(String metric) {
+        Set allowed_metrics = ModelMetrics.getAllowedMetrics(_model._key);
+        metric = metric.toLowerCase();
+        if (metric.equals("auto")) {
+            if (_model._output._training_metrics instanceof ModelMetricsBinomial)
+                metric = "auc";
+            else if (_model._output._training_metrics instanceof ModelMetricsRegression)
+                metric = "rmse";
+            else if (_model._output._training_metrics instanceof ModelMetricsMultinomial)
+                metric = "logloss";
+            else
+                throw new IllegalArgumentException("Unable to infer metric. Please specify metric for permutation variable importance.");
+        }
+
+        if (!allowed_metrics.contains(metric))
+            throw new IllegalArgumentException("Permutation Variable Importance doesn't support " + metric + " for model " + _model._key);
+
+        return metric;
+    }
+
+
+    private HashMap<String, Double> calculatePermutationVarImp(String metric, long n_samples, final String[] features, long seed){
+        // Use random seed if set to -1
+        if (-1 == seed) seed = new Random().nextLong();
+
+        if (n_samples == 1)
+            throw new IllegalArgumentException("Unable to permute one row. Please set n_samples to higher value or to -1 to use the whole dataset.");
+
+        final String[] variables = _inputFrame.names();
+
+        HashSet<String> featuresToCompute = new HashSet<>(Arrays.asList((null != features && features.length > 0) ? features : variables));
+        featuresToCompute.removeAll(Arrays.asList(_model._parms.getNonPredictors()));
+        if (_model._parms._ignored_columns != null)
+            featuresToCompute.removeAll(Arrays.asList(_model._parms._ignored_columns));
+
+        Frame fr = null;
+        if (n_samples > 1) {
+            fr = MRUtils.sampleFrame(_inputFrame, n_samples, _model._parms._weights_column, seed);
+            while (fr.numRows() < 2) {
```

Comment:
hmm, you could just generate n row indices and extract those from the frame using Frame#deepSlice

so for low numbers you could do deepSlice and for big data use the approximate approach

Reply:
Good idea, thanks! Is the threshold for low numbers 1000 ok? (`n_samples < 1000` use deepSlice otherwise the sampleFrame)